Resolves Bless/Curse and Sealed tokens

dscarpac opened the pull request:
This PR changes add several features:

(1) Allow for cards to resolve tokens sealed on them as if they came from the bag (used via context menu). As an example, you could use Favor of the Sun to resolve a Bless token, which moves it to the normal area where tokens are revealed from the bag, and then right click on the draw chaos token button to continue resolving your test without the Bless token going back into the bag.

(2) Allow for cards to resolve specific tokens from the bag instead of a random one. Currently this functionality is available on Kohaku's Book of Living Myths, with an XML helper enabled via context menu:

![image](https://github.com/argonui/SCED/assets/148364064/f49e6815-40fd-4127-8206-e5359d831c7b)

The helper also only lets you whichever token (Bless or Curse) there is more of in the bag, as per the rule text on the card. Again, this allows for additional tokens to be revealed without breaking up the "reveal tokens chain" in the play area. Forthcoming will be similar functionality for pulling Nkosi Mabati's "sigil" out of the bag automatically.

Chr1Z93 commented on objects/AllPlayerCards.15bb07/BookofLivingMyths.c5fb1f.json:
```diff
@@ -58,5 +58,23 @@
+      "Name": "curse",
+      "URL": "http://cloud-3.steamusercontent.com/ugc/2380784374775547135/2360372CBE9452CB7B4D135BE13BBA6D46B7D427/"
+    }
+    ]
+
```
Nitpick: remove empty line and fix indentation for the final bracket (or this will show up as change with the next unbundling)

Chr1Z93 commented on objects/AllPlayerCards.15bb07/NkosiMabati3.6c5628.json:
What's changed here?

Chr1Z93 commented on src/chaosbag/ChaosBagApi.ttslua:
```diff
@@ -69,8 +69,8 @@ do
-  ChaosBagApi.drawChaosToken = function(mat, drawAdditional)
-    return Global.call("drawChaosToken", {mat = mat, drawAdditional = drawAdditional})
+  ChaosBagApi.drawChaosToken = function(mat, drawAdditional, tokenType, cardSealedToken)
+    return Global.call("drawChaosToken", {mat = mat, drawAdditional = drawAdditional, tokenType = tokenType, sealedToken = guidToBeResolved})
```
Add documentation for the new parameters (like for the others above)

Chr1Z93 commented on src/playercards/cards/FavoroftheSun1.ttslua:
```diff
@@ -6,3 +6,26 @@ SHOW_SINGLE_RELEASE = true
+    local mat = guidReferenceApi.getObjectByOwnerAndType(closestMatColor, "Playermat")
+    guidToBeResolved = table.remove(sealedTokens)
+    chaosBagApi.drawChaosToken(mat, true, _, guidToBeResolved)
+end
```
Since this code is identical for the Favors and the Codex, it would be better if it is put somewhere else (either in a separate file or part of the library file - in the latter case this would be activated by some variable in this script like the other things)

Chr1Z93 commented on xml/playercards/BookofLivingMyths.xml:
While this is perfectly fine for this case, this XML should be added via scripting if more cards get this feature (to allow dynamically setting parts of it).

Chr1Z93 commented on src/core/Global.ttslua:
```diff
@@ -303,12 +303,36 @@ function drawChaosToken(params)
+              end
+          end
+      else
+        tokenIndex = 0
```
This doesn't do anything since it starts of as 0

Chr1Z93 commented on src/core/Global.ttslua:
```diff
@@ -303,12 +303,36 @@ function drawChaosToken(params)
+      token = getObjectFromGUID(params.sealedToken)
+      token.setPositionSmooth(params.mat.positionToWorld(tokenOffset))
+      local guid = token.getGUID()
+      local type = token.getName()
```
Nitpick: This variable should not be called type (either `name` or `tokenType` would be better imo)

Chr1Z93 commented on src/core/Global.ttslua:
```diff
@@ -303,12 +303,36 @@ function drawChaosToken(params)
+      end
 
+      token = chaosBag.takeObject({
+        index = tokenIndex,
```
It might not actually make a difference, but I'd rather avoid specifying an index if not looking for a specific token.

Entrox-Licher reviewed (changes requested):
Excellent functionality!

A few functional change requests:

- Change context menu fail to find broadcast to yellow instead of red (not really an error)
- Give a feedback when the 'Resolve Token' buttons fail to find, just like you do with the context menu option
- Change the background of the 'Resolve Curse' button.  I see what you were going for, but currently it looks grayed out like you can't use it.  I'd use for distinctive colors or just keep them the same.

dscarpac commented:
Actually, in that image the curse button is supposed to be grayed out because you indeed can't use it! I should have included images for all of the possibilities. Here's with both active.
![image](https://github.com/argonui/SCED/assets/148364064/6c96bd97-c530-4e72-8dd5-bc3df982581f)

dscarpac commented:
> Excellent functionality!
> 
> A few functional change requests:
> 
> * Change context menu fail to find broadcast to yellow instead of red (not really an error)
> * Give a feedback when the 'Resolve Token' buttons fail to find, just like you do with the context menu option
> * Change the background of the 'Resolve Curse' button.  I see what you were going for, but currently it looks grayed out like you can't use it.  I'd use for distinctive colors or just keep them the same.

I addressed all of these in the new update -- except for the red broadcast color, I think we need documentation about what colors mean what somewhere first. I also changed the colors to this:
![image](https://github.com/argonui/SCED/assets/148364064/28a94147-f8ce-454b-b57c-6846b01a4967)

Chr1Z93 commented on src/playercards/CardsThatSealTokens.ttslua:
```diff
@@ -76,8 +76,10 @@ Thus it should be implemented like this:
 local blessCurseManagerApi = require("chaosbag/BlessCurseManagerApi")
 local chaosBagApi = require("chaosbag/ChaosBagApi")
 local tokenArrangerApi = require("accessories/TokenArrangerApi")
+local playmatApi = require("playermat/PlaymatApi")
```
This should be above "tokenArrangerApi" to keep things alphabetically sorted

Chr1Z93 commented on src/playercards/CardsThatSealTokens.ttslua:
```diff
@@ -76,8 +76,10 @@ Thus it should be implemented like this:
 local chaosBagApi = require("chaosbag/ChaosBagApi")
 local tokenArrangerApi = require("accessories/TokenArrangerApi")
+local playmatApi = require("playermat/PlaymatApi")
+local guidReferenceApi = require("core/GUIDReferenceApi")
```
see above (keep require's alphabetically sorted)

Chr1Z93 commented on src/playercards/CardsThatSealTokens.ttslua:
```diff
@@ -262,3 +273,15 @@ function returnToken(guid)
+  local mat = guidReferenceApi.getObjectByOwnerAndType(closestMatColor, "Playermat")
+  guidToBeResolved = table.remove(sealedTokens)
+  chaosBagApi.drawChaosToken(mat, true, _, guidToBeResolved)
+end
```
add missing end of file newline

Chr1Z93 commented on src/playercards/cards/BookofLivingMyths.ttslua:
```diff
@@ -102,4 +104,14 @@ function countBlessCurse()
+function errorMessage ()
+  if currentState == "Empty" then
+    broadcastToAll("There are no Bless or Curse tokens in the chaos bag.","Red")
+  elseif currentState =="More Bless" then
```
Nitpick: missing space after the "=="

Chr1Z93 commented on src/playercards/cards/BookofLivingMyths.ttslua:
```diff
@@ -102,4 +104,14 @@ function countBlessCurse()
+    broadcastToAll("There are more Bless tokens than Curse tokens in the chaos bag.","Red")
+  else
+    broadcastToAll("There are more Curse tokens than Bless tokens in the chaos bag.","Red")
+  end
```
Missing end of file new line

Chr1Z93 commented on src/core/Global.ttslua:
```diff
@@ -302,12 +302,34 @@ function drawChaosToken(params)
-      position = params.mat.positionToWorld(tokenOffset),
-      rotation = params.mat.getRotation()
-    })
+        
```
There should be `local token` at this point (otherwise the code below will set the global `token`)

Chr1Z93 commented on src/core/Global.ttslua:
```diff
@@ -302,12 +302,34 @@ function drawChaosToken(params)
-    })
+        
+    if params.sealedToken then -- resolve a sealed token from a card
+     
```
Excess empty line

Chr1Z93 commented on src/core/Global.ttslua:
```diff
@@ -302,12 +302,34 @@ function drawChaosToken(params)
+      end
 
+      token = chaosBag.takeObject(takeParameters)
+      
```
Excess empty line

Chr1Z93 commented on src/playercards/CardsThatSealTokens.ttslua:
```diff
@@ -262,3 +273,15 @@ function returnToken(guid)
+  end
+  local closestMatColor = playmatApi.getMatColorByPosition(self.getPosition())
+  local mat = guidReferenceApi.getObjectByOwnerAndType(closestMatColor, "Playermat")
+  guidToBeResolved = table.remove(sealedTokens)
```
Should be local variable

Chr1Z93 commented on src/playercards/cards/BookofLivingMyths.ttslua:
```diff
@@ -0,0 +1,107 @@
+local playmatApi = require("playermat/PlaymatApi")
+local guidReferenceApi = require("core/GUIDReferenceApi")
+local chaosBagApi      = require("chaosbag/ChaosBagApi")
```
Sort `require` statements alphabetically (and either align all `=` or none)

Chr1Z93 commented on src/playercards/cards/BookofLivingMyths.ttslua:
```diff
@@ -0,0 +1,107 @@
+local guidReferenceApi = require("core/GUIDReferenceApi")
+local chaosBagApi      = require("chaosbag/ChaosBagApi")
+
+
```
Excess empty line

Chr1Z93 commented on src/playercards/cards/BookofLivingMyths.ttslua:
```diff
@@ -0,0 +1,107 @@
+  self.addContextMenuItem("Enable Helper", createButtons)
+  if savedData ~= "" then
+    local loadedData = JSON.decode(savedData)
+    log(loadedData)
```
Remove `log()` statement

Chr1Z93 commented on src/playercards/cards/BookofLivingMyths.ttslua:
```diff
@@ -0,0 +1,107 @@
+
+function countBlessCurse()
+-- count tokens in the bag and show appropriate buttons
+  numInBag = { Bless = 0, Curse = 0 }
```
should be local variable

Chr1Z93 commented on src/playercards/cards/BookofLivingMyths.ttslua:
```diff
@@ -0,0 +1,107 @@
+end
+
+function countBlessCurse()
+-- count tokens in the bag and show appropriate buttons
```
comment should be above the function declaration

Chr1Z93 commented on src/playercards/cards/FavoroftheMoon1.ttslua:
```diff
@@ -4,5 +4,6 @@ VALID_TOKENS = {
+RESOLVE_TOKEN = true
 
-require("playercards/CardsThatSealTokens")
+require("playercards/CardsThatSealTokens")
```
Missing end of file newline

Chr1Z93 commented on src/playercards/cards/FavoroftheSun1.ttslua:
```diff
@@ -4,5 +4,6 @@ VALID_TOKENS = {
+RESOLVE_TOKEN = true
 
-require("playercards/CardsThatSealTokens")
+require("playercards/CardsThatSealTokens")
```
Missing end of file newline

Chr1Z93 commented on src/playercards/cards/TheCodexofAges.ttslua:
```diff
@@ -2,4 +2,6 @@ VALID_TOKENS = {
-require("playercards/CardsThatSealTokens")
+RESOLVE_TOKEN = true
+
+require("playercards/CardsThatSealTokens")
```
Missing end of file newline

Chr1Z93 commented on xml/playercards/BookofLivingMyths.xml:
```diff
@@ -0,0 +1,41 @@
+      </Cell>
+    </Row>
+  </TableLayout>
+</Panel>
```
Missing end of file newline

Chr1Z93 commented on xml/playercards/BookofLivingMyths.xml:
If you wanna get fancy, you can slim this XML down by using defaults (example for the buttons):
```xml
<Defaults>
  <Button padding="50 50 50 50"
    font="font_teutonic-arkham"
    fontSize="300"
    iconAlignment="Right"/>
</Defaults>
```
This will automatically apply to all <Button> elements. Check out the OptionPanel.xml for a bigger example.

Chr1Z93 reviewed (changes requested):
Fully functional, this is just super nitpicky stuff
